lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -1472,81 +1472,97 @@ bool MergeMatMul::run(Function *F, const CompilationContext &cctx) {
-    // Pad and Convolution. The 'pads' array won't have the same size because
-    // only spatial dimensions are specified for the convolution while all
-    // dimensions are specified for Pad.
+  // Compute the new padding.
+  // Note: - convolution only supports positive padding
+  //       - the convolution takes NHWC input tensors.
+  bool canMerge = true;
+  auto padPads = PN->getPads();
+  auto convPads = CN->getPads();
 
-    // The merge can apply only if no padding is requested for non spatial
-    // dimensions.
-    if ((padPads[0] != 0) || (padPads[3] != 0) || (padPads[4] != 0) ||
-        (padPads[7] != 0)) {
-      continue;
-    }
+  // For now, there is a different interpretation of the ONNX spec for
+  // Pad and Convolution. The 'pads' array won't have the same size because
+  // only spatial dimensions are specified for the convolution while all
+  // dimensions are specified for Pad.
 
-    // Compute new spatial padding.
-    const int H_INDEX = 1;
-    std::vector<unsigned_t> newConvPads(4);
-    auto numDims = PN->getResult().dims().size();
-    for (size_t i = 0; i < 2; i++) {
-      // Two pad integers per dimension (begin and end padding).
-      for (size_t j = 0; j < 2; j++) {
-        int newConvPadSigned =
-            padPads[(i + H_INDEX) + j * numDims] + int(convPads[i + j * 2]);
-        if (newConvPadSigned < 0) {
-          canMerge = false;
-          break;
-        }
-        newConvPads[i + j * 2] = unsigned_t(newConvPadSigned);
+  // The merge can apply only if no padding is requested for non spatial
+  // dimensions.
+  if ((padPads[0] != 0) || (padPads[3] != 0) || (padPads[4] != 0) ||
+      (padPads[7] != 0)) {
+    return false;
+  }
+
+  // Compute new spatial padding.
+  const int H_INDEX = 1;
+  std::vector<unsigned_t> newConvPads(4);
+  auto numDims = PN->getResult().dims().size();
+  for (size_t i = 0; i < 2; i++) {
+    // Two pad integers per dimension (begin and end padding).
+    for (size_t j = 0; j < 2; j++) {
+      int newConvPadSigned =
+          padPads[(i + H_INDEX) + j * numDims] + int(convPads[i + j * 2]);
+      if (newConvPadSigned < 0) {
+        canMerge = false;
+        break;
       }
+      newConvPads[i + j * 2] = unsigned_t(newConvPadSigned);
     }
-    if (!canMerge) {
-      continue;
-    }
+  }
+  if (!canMerge) {
+    return false;
+  }
 
-    // New Convolution
-    auto *newCN = F->createConv(CN->getName(), PN->getInput(), CN->getFilter(),
-                                CN->getBias(), CN->getResult().getType(),
-                                CN->getKernels(), CN->getStrides(), newConvPads,
-                                CN->getGroup(), CN->getDilation());
-    newCN->setFusedActivation(CN->getFusedActivation());
-    newCN->setFusedActivationArgs(CN->getFusedActivationArgs());
+  // Set new pads and remove PadNode.
+  CN->setPads(newConvPads);
+  PN->getResult().typeUnsafeReplaceAllUsesOfWith(PN->getInput());
```

Comment:
So we now need to do an unsafe replace here. Previously we were changing the pads and passing it into a new Conv node `newCN`, why isn't it possible to keep the same type output for the replacement once we merge the padding in? Doesn't this mean that `CN` has an incorrect `outTy` after we call `setPads` on it? Can you explain a bit more/if we need to keep it add into a comment here why this is necessary?

Reply:
The logic is actually correct:
- **previous logic**: create new conv node between the Pad input and old conv output
- **new logic**: modify current conv node parameters and remove pad node.

So the line of code you showed actually removes the Pad node which of course has different types for its input and output.

Reply:
This line indeed looks confusing. Would `CV->setInput(PN->getInput())` work instead? I believe this would also allow to avoid Pad single user limitation (see my comment above).